http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoMetricsMonitor.scala

```diff
@@ -0,0 +1,342 @@
+    appHealth <- countAppsByHealth(clusters)
+    _ <- recordMetric(appHealth)
+    _ <- logger.info(s"Recorded health metrics for ${appHealth.size} apps")
+  } yield ()
+
+  /** Queries the DB for all active runtimes and collects metrics */
+  private[monitor] def retrieveRuntimeMetrics(implicit ev: Ask[F, AppContext]): F[Unit] = for {
+    runtimeSeq <- clusterQuery.listActiveForMetrics.transaction
+    runtimes = runtimeSeq.toList
+    runtimeDbStatus = countRuntimesByDbStatus(runtimes)
+    _ <- recordMetric(runtimeDbStatus)
+    _ <- logger.info(s"Recorded status metrics for ${runtimeDbStatus.size} runtimes")
+    runtimeHealth <- countRuntimesByHealth(runtimes)
+    _ <- recordMetric(runtimeHealth)
+    _ <- logger.info(s"Recorded health metrics for ${runtimeHealth.size} runtimes")
+  } yield ()
+
+  /** Transforms DB apps to AppStatusMetric and computes counts. */
+  private[monitor] def countAppsByDbStatus(
+    allClusters: List[KubernetesCluster]
+  ): Map[AppStatusMetric, Int] = {
+    val allApps = for {
+      c <- allClusters
+      n <- c.nodepools
+      a <- n.apps
+    } yield Map(
+      AppStatusMetric(c.cloudContext.cloudProvider,
+                      a.appType,
+                      a.status,
+                      getRuntimeUI(a.labels),
+                      getAzureCloudContext(c.cloudContext)
+      ) -> 1
+    )
+
+    allApps.combineAll
+  }
+
+  /** Transforms DB runtimes to RuntimeStatusMetric and computes counts. */
+  private[monitor] def countRuntimesByDbStatus(allRuntimes: List[RuntimeMetrics]): Map[RuntimeStatusMetric, Int] = {
+    val allContainers = for {
+      r <- allRuntimes
+      c <- r.containers if c != RuntimeContainerServiceType.WelderService
+    } yield Map(
+      RuntimeStatusMetric(r.cloudContext.cloudProvider,
+                          c.imageType,
+                          r.status,
+                          getRuntimeUI(r.labels),
+                          getAzureCloudContext(r.cloudContext)
+      ) -> 1
+    )
+
+    allContainers.combineAll
+  }
+
+  /** Performs health checks for Running apps, and transforms to AppHealthMetric. */
+  private[monitor] def countAppsByHealth(
+    allClusters: List[KubernetesCluster]
+  )(implicit ev: Ask[F, AppContext]): F[Map[AppHealthMetric, Int]] = {
+    val allServices = for {
+      c <- allClusters if c.asyncFields.isDefined
+      n <- c.nodepools
+      // Only care about Running apps for health check metrics
+      a <- n.apps if a.status == AppStatus.Running
+      s <- a.appResources.services
+    } yield (c.cloudContext, c.asyncFields.get.loadBalancerIp, a, s.config.name)
+
+    allServices
+      .parTraverseN(parallelism) { case (cloudContext, baseUri, app, serviceName) =>
+        for {
+          ctx <- ev.ask
+          // For GCP just test if the app is available through the Leo proxy.
+          // For Azure impersonate the user and call the app's status endpoint via Azure Relay.
```

Comment:
Oh whoa, naively this looks a lot more complex / expensie to do

Reply:
It is! Mostly because of the Azure relay setup, as opposed to the Leo proxy. Perhaps we could have some utilities in code to simplify this.